### PR TITLE
A mish-mash of micro-optimizations

### PR DESCRIPTION
**Pull request by nnethercote**
These were aimed at speeding up LLVM codegen, but ended up affecting other places as well.

r? @bjorn3

### CODE REVIEW

**Comment by oli-obk** [comment]
@rust-timer build aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513

**Comment by nnethercote** [comment]
The walltime/cycles improvements of up to 8% for `tt-muncher` look spurious; I think it has just become noisy on those metrics. Everything else should be real, though. I think the last commit ("Avoid cloning `LocalDecls`) is the most effective change.

File: compiler/rustc_codegen_llvm/src/builder.rs [hunk 1]
```diff
@@ -225,17 +226,21 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
         let args = self.check_call("invoke", llty, llfn, args);
         let funclet_bundle = funclet.map(|funclet| funclet.bundle());
         let funclet_bundle = funclet_bundle.as_ref().map(|b| &*b.raw);
-        let mut bundles = vec![funclet_bundle];
+        let mut bundles: SmallVec<[_; 2]> = SmallVec::new();
+        if funclet_bundle.is_some() {
+            bundles.push(funclet_bundle);
```
Comment (oli-obk) [inline compiler/rustc_codegen_llvm/src/builder.rs]: maybe we could stop putting options in there and just use the value directly in LLVMRustBuildCall?

Comment (oli-obk) [reply compiler/rustc_codegen_llvm/src/builder.rs]: 🤣 I should have looked at the next commit

**Review by oli-obk — APPROVED** [review]

**Pull request merged**